ompi/datatype/ompi_datatype_lookup_by_opal_id.c

```diff
@@ -0,0 +1,28 @@
+/* -*- Mode: C; c-basic-offset:4 ; -*- */
+/*
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+
+#include "ompi_config.h"
+#include "ompi/datatype/ompi_datatype.h"
+#include "ompi/datatype/ompi_datatype_internal.h"
+
+
+ompi_datatype_t* ompi_datatype_lookup_by_opal_id( uint16_t opal_id )
```

Comment:
OPAL and OMPI datatypes are different things, and multiple OMPI datatypes will be mapped unto the same underlying OPAL datatype. Thus, looking for an OMPI datatype that matches a specific OPAL id, will give you the first OMPI datatype that matches that id, and that might be an OMPI datatype that is equivalent to your original one, but not necessarily the same. Why are you not always manipulating OMPI datatype id (which would be unique for each datatype).

Reply:
@bosilca I see what you are saying.  The underlying event support is implemented in opal/mca/base/mca_base_event.* with opal_datatype_t, which is a different size than ompi_datatype_t (uint16_t / uint32_t).  The currently implemented events are in OMPI modules.  Do you have a suggestion for tracking OMPI datatypes with the OPAL support?

Reply:
I think the intent here is:

* OPAL provides a basic datatype engine.  It's not specifically tied to the MPI API or MPI semantics.  The OPAL datatype engine does generic datatype-like things.
* The datatype functionality in MPI layer adds further semantics and data structures on top of the OPAL datatype engine to implement the MPI-specific datatype functionality.

I think @bosilca is asking to maintain the clean separation between the two.  Hence, you'll need to look up MPI datatypes as MPI datatypes -- not OPAL datatypes.

The same concept applies to the base MPI_T event support: it should be tracking MPI events and MPI datatypes and other MPI objects.  If desirable, there could be an underlying OPAL "event tracking" engine that does things in an MPI-agnostic way (wholly independent of the MPI layer APIs and MPI semantics).  This would be useful, for example, if we want to be able to track OPAL-level events that have no knowledge of the MPI layer (the BTLs seem like an obvious candidate here, for example).

The MPI_T functionality in the MPI layer should be able to call down into the OPAL "event tracking" engine to track MPI-level events.

The tricky part will then be how to correlate MPI-level events with OPAL-level events (e.g., how to correlate BTL events to specific MPI communicators, for example).